### PR TITLE
[WIP] ci: support regression tests on all platforms

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# Code Review Checklist

* [x] Support periodical `llvm-lit` regression tests on all supported platforms.
* [x] Disable unsupported tests on Windows
* [x] Add Slack notification if periodic run fails

## Tests

[Testing workflow](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704)

## Purpose

Periodically test on MacOS and Windows to confirm there are no regressions as well as on Linux.

### CODE REVIEW

**Comment by FlashSheridan** [comment]
This may be more of a question for the GitHUB UI designer than for you, but how do we find the tests run on Mac and Windows?  All six of the checks above seem to have Ubuntu as their Operating System near the beginning of the log, and none seem to show the Matrix: Build, with Mac ARM & x86, and Windows, as in e.g. [runs/9841084704](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704)/Show All Jobs.

**Comment by FlashSheridan** [comment]
Another UI nit:  “Windows” should probably display as something like “Windows-x86” (assuming that’s what that run is), in deference to the future expansion of Windows on ARM.  This would also be more consistent with Linux and Mac.

File: llvm/test/tools/llvm-cov/native_separators.c
Comment (FlashSheridan) [inline llvm/test/tools/llvm-cov/native_separators.c:8]: This and the following lines are an odd combination, especially give the opening comment.  More generally, I would have liked comments on at least some of the Windows Unsupported directives.  Even more generally, some indication of which are because of LLVM’s relative lack of interest in Windows, and which because of Windows’ many limitations and nonconformities.  (Both can be true, of course.)

File: .github/workflows/build-binaries.yml
Comment (FlashSheridan) [inline .github/workflows/build-binaries.yml]: To be fair to the preceding test, I think this should be named something like “Running Lit tests (Non-Windows),” or, depending on one’s feelings about Windows, “(Normal)” or “(Non-broken).

File: llvm/test/tools/llvm-cov/native_separators.c
Comment (antonbaliasnikov) [reply llvm/test/tools/llvm-cov/native_separators.c:8]: @FlashSheridan , thanks a lot for a thorough review. You're absolutely right that we need to categorize and add more comments in all of these cases.

I actually wanted to separate these issues into categories and understand on how to proceed further with them.

LLVM issues can be split into the following categories:

1. Windows MSYS CI environment returns `BAD-DATE-FORMAT` instead of the date and it fails the following tests that checks the date in output:
`LLVM :: Object/archive-big-read.test`
`LLVM :: Object/archive-toc.test`
`LLVM :: Object/archive-extract.test`
`LLVM :: tools/llvm-ar/empty-uid-gid.test`
`LLVM :: tools/llvm-libtool-darwin/deterministic-library.test`
`LLVM :: tools/llvm-objcopy/ELF/deterministic-archive.test`
`LLVM :: tools/llvm-ranlib/D-flag.test`

2. "Unsupported file type" string returned in our Windows MSYS environment instead of "malformed file" that is expected:
`LLVM :: tools/llvm-nm/invalid-tapi-files.test`
`LLVM :: tools/llvm-nm/tapi-files.test`
`LLVM :: tools/llvm-tapi-diff/v5.test`

3. Directory and Linux/Windows slash `/``\` issues:
* `LLVM :: TableGen/x86-fold-tables.td` - Linux specific slashes in test, not working on Windows properly
* `LLVM :: tools/llvm-cov/native_separators.c` - Windows-specific test, but it will never work in msys environment as it tries to use \tmp directory. It will only work on Windows native.
* `LLVM :: tools/llvm-cov/relative-dir.test` - multiple slashes on MSys are generated and leads to wrong output check
* `LLVM-Unit :: Support/./SupportTests.exe/55/83` - path issue that is impossible to debug without native Windows environment

I'm not sure what to do with that, it's impossible to debug without native and MSYS environments together, and this can be some bug in the LLVM Support Chrono library. These tests are also not really relevant for us, so I believe it's probably OK to disable them completely marking them as `UNSUPPORTED` on Windows and adding a comment about it. What do you think?

The next thing is the LLD tests:

1. Require proper investigation when we have Windows native host:
* `LLD-Unit :: AsLibELF/./LLDAsLibELFTests.exe/1/2` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-code-reloc.s` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-data-reloc.s`- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-errors.s` - Some additional output "ld.lld: warning: cannot find entry symbol _start; not setting start address" that is not expected by the test
* `lld :: ELF/invalid/symtab-symbols.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eh-frame-type.test` - Non zero exit code returned by ld.lld
* `lld :: ELF/llvm33-rela-outside-group.s` - Exception fail in llvm::PassRegistry::~PassRegistry()
* `lld :: ELF/relocation-group.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/relocation-none.test `- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/section-align-0.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/reloc-sec-before-relocated.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/relocation-rel-format.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/sht-group-empty.test `- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/shf-info-link.test` - Non zero exit code returned by ld.lld without any stderr output

2. MacOS specific tests that are unsupported for Windows by default:
`lld :: MachO/error-limit.test`
`lld :: MachO/double-unwind-info.s`
`lld :: MachO/local-private-extern.yaml`

For these issues, I've created [CPR-1784](https://linear.app/matterlabs/issue/CPR-1784/llvm-check-lld-regression-tests-windows-failures) and they will require investigation as soon as we have Windows host.

I was also thinking to disable them for now, because I was unable to make them to `--xfail` properly, it's just doesn't work on Windows :(

It's a huge pain to investigate and debug these issues without a Windows host.

**Comment by antonbaliasnikov** [comment]
> This may be more of a question for the GitHUB UI designer than for you, but how do we find the tests run on Mac and Windows? All six of the checks above seem to have Ubuntu as their Operating System near the beginning of the log, and none seem to show the Matrix: Build, with Mac ARM & x86, and Windows, as in e.g. [runs/9841084704](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704)/Show All Jobs.

@FlashSheridan , could you, please, point me to the place where you see `Ubuntu` in the logs? When I'm opening an individual job, I see only the host specific output. Maybe, I'm looking in the wrong place.

**Comment by antonbaliasnikov** [comment]
> Another UI nit: “Windows” should probably display as something like “Windows-x86” (assuming that’s what that run is), in deference to the future expansion of Windows on ARM. This would also be more consistent with Linux and Mac.

Yeah, we can do that enhancement, I'll update it.

**Comment by FlashSheridan** [comment]
> > This may be more of a question for the GitHUB UI designer than for you, but how do we find the tests run on Mac and Windows? All six of the checks above seem to have Ubuntu as their Operating System near the beginning of the log, and **none seem to show the Matrix: Build**, with Mac ARM & x86, and Windows, as in e.g. [runs/9841084704](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704)/Show All Jobs.
> 
> @FlashSheridan , could you, please, point me to the place where you see `Ubuntu` in the logs? When I'm opening an individual job, I see only the host specific output. Maybe, I'm looking in the wrong place.

My complaint about the GitHub UI may no longer be relevant, either because the tests have now completed and display well, or user error on my part; my link above now _does_ “show the Matrix: Build.”  From there it’s evident how to navigate to, e.g. “Operating System  /  Microsoft Windows Server 2022”

But I still find the GitHub UI baffling:  When I cmd-check on (now) “All checks have passed” / “Show all checks” / e.g. the last check, “Regression tests” / “Details”; then (default “regression-tests”) on “Set up job,” the machine name is “Ubuntu-2204-jammy-amd64-base.”  “Prepare-test-matrix” / “Set up job” / Operating System says “Ubuntu.” 

Documenting (perhaps just for me and future newbies) where to click (and/or not to) may be all that’s needed.

File: llvm/test/tools/llvm-cov/native_separators.c
Comment (FlashSheridan) [reply llvm/test/tools/llvm-cov/native_separators.c:8]: _(1 of 2)_
> It's a huge pain to investigate and debug these issues without a Windows host.

Of course; supporting Windows (or any platform) has an absolute requirement for at least one machine (perhaps virtual) to test it on; I’ve made that clear in planning documents since before you started at Matter Labs.  I don’t think it’s a good use of your time to waste it on unnecessarily difficult experimentation while waiting for required hardware.  If you need backing for escalating this, count me in, though given the cheapness of Windows hardware, it seems to be covered by policy…

Comment (FlashSheridan) [reply llvm/test/tools/llvm-cov/native_separators.c:8]: _(2 of 2)_

> You're absolutely right that we need to categorize and add more comments in all of these cases.
> I actually wanted to separate these issues into categories and understand on how to proceed further with them.
> LLVM issues can be split into the following categories: …

This is excellent work and deserves a wider audience, both internal and upstream, even if it exceeds in both length and quality what’s customary for LLVM’s Windows-specific documentation and comments.  If I were working on another LLVM fork and struggling with Windows support, I’d be delighted to see your comments (at the latest when we upstream our fork, perhaps earlier) for each suppression, the information you’ve elicited on whether the suppression is because of:
1.  a genuine Windows limitation, 
2. legitimate target-specificity in the test
3. laziness in the test, 
4. a genuine bug in the upstream software under test, 
5. a (preferably temporary) limitation in our code, or
6. currently unknown.

**Comment by antonbaliasnikov** [comment]
> > > This may be more of a question for the GitHUB UI designer than for you, but how do we find the tests run on Mac and Windows? All six of the checks above seem to have Ubuntu as their Operating System near the beginning of the log, and **none seem to show the Matrix: Build**, with Mac ARM & x86, and Windows, as in e.g. [runs/9841084704](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704)/Show All Jobs.
> > 
> > 
> > @FlashSheridan , could you, please, point me to the place where you see `Ubuntu` in the logs? When I'm opening an individual job, I see only the host specific output. Maybe, I'm looking in the wrong place.
> 
> My complaint about the GitHub UI may no longer be relevant, either because the tests have now completed and display well, or user error on my part; my link above now _does_ “show the Matrix: Build.” From there it’s evident how to navigate to, e.g. “Operating System / Microsoft Windows Server 2022”
> 
> But I still find the GitHub UI baffling: When I cmd-check on (now) “All checks have passed” / “Show all checks” / e.g. the last check, “Regression tests” / “Details”; then (default “regression-tests”) on “Set up job,” the machine name is “Ubuntu-2204-jammy-amd64-base.” “Prepare-test-matrix” / “Set up job” / Operating System says “Ubuntu.”
> 
> Documenting (perhaps just for me and future newbies) where to click (and/or not to) may be all that’s needed.

Oh, I understood. The thing is that the [testing workflow](https://github.com/matter-labs/era-compiler-llvm/actions/runs/9841084704) I've shared is a separate manually-triggered workflow, and it's not displayed in the "Checks" for the pull request. "Checks" include only Linux x86, and this is what you see.

We will be running Windows and MacOS periodically - right now, it is set to run monthly (or manually, by request).

We can think of including other platforms in Pull Request's regressions too, I believe. But for now, I was implementing it as per your requirements in CPR-1397: _"This presumably won’t run mandatorily for each PR push, so it doesn’t need to be especially fast, but should run automatically periodically."_

**Comment by antonbaliasnikov** [comment]
@FlashSheridan , I've addressed all issues and added comments for each test. Please review once again, and I'll merge it after your approval.

**Comment by FlashSheridan** [comment]
> @FlashSheridan , I've addressed all issues and added comments for each test. Please review once again, and I'll merge it after your approval.

Thanks, I’ll start looking soon.  (I would have appreciated a non-force push so I could review the diffs…)

**Comment by antonbaliasnikov** [comment]
> > @FlashSheridan , I've addressed all issues and added comments for each test. Please review once again, and I'll merge it after your approval.
> 
> Thanks, I’ll start looking soon. (I would have appreciated a non-force push so I could review the diffs…)

@FlashSheridan , sorry for the force push. We actually have a strategy that checks each commit, and we always have to squash and force push to test correctly and make the history cleaner before we merge.

If I do an iterative fix, I'll have to run CI twice, test unnecessary commits, and then anyway squash before merge which will put a successful review to stale.

So, I believe with the approach we have in the LLVM repo, we cannot review the diffs properly :(

**Comment by FlashSheridan** [comment]
> @FlashSheridan , sorry for the force push. We actually have a strategy that checks each commit, and we always have to squash and force push to test correctly and make the history cleaner before we merge.
> If I do an iterative fix, I'll have to run CI twice, test unnecessary commits, and then anyway squash before merge which will put a successful review to stale.
> So, I believe with the approach we have in the LLVM repo, we cannot review the diffs properly :(

Ouch, understood (and I’ve resigned myself to “then anyway squash before merge”), but it does mean that I can’t, for instance, verify your (presumably trivial) fix

> On Jul 10, 2024, at 7:11 AM, Anton Baliasnikov <notifications@github.com> wrote:
> [@antonbaliasnikov](https://github.com/antonbaliasnikov) pushed 1 commit.
> 	•	[25c45d8](https://github.com/matter-labs/era-compiler-llvm/commit/25c45d89516b690975c8e48982570470b233b60d) ci: support regression tests on all platforms
>

presumably for:

On Jul 10, 2024, at 6:54 AM, github-actions[bot] <notifications@github.com> wrote:
> 
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> … View the diff from clang-format here.
> diff --git a/llvm/unittests/Support/CommandLineTest.cpp b/llvm/unittests/Support/CommandLineTest.cpp
> index 8ccf46888..11dc0f2c2 100644
> --- a/llvm/unittests/Support/CommandLineTest.cpp
> +++ b/llvm/unittests/Support/CommandLineTest.cpp
> @@ -881,7 +881,8 @@ TEST(CommandLineTest, ResponseFiles) {
>  }
>  
>  TEST(CommandLineTest, RecursiveResponseFiles) {
> -  // Temporary disable on Windows due to issues with backslashes in paths on MSYS2.
> +  // Temporary disable on Windows due to issues with backslashes in paths on
> +  // MSYS2.
…

File: lld/test/ELF/eh-frame-type.test
Comment (FlashSheridan) [inline lld/test/ELF/eh-frame-type.test:3]: This file, unlike the following ones, seems to exist upstream; so your observation may be a valuable contribution to the LLVM community.  (Though the maintainers might not react diligently to a documentation PR.)  OTOH, it might be something specific to us.  So I think that either this comment should be wrapped in Matter Labs’ Begin/End markers, and/or it should be upstreamed so that we don’t have to do that.

File: lld/test/ELF/invalid/symtab-symbols.test
Comment (FlashSheridan) [inline lld/test/ELF/invalid/symtab-symbols.test:1]: Similar query to that above for lld/test/ELF/eh-frame-type.test.  Likewise, I suspect from the names, for the following few files.

File: lld/test/ELF/section-symbols.test
Comment (FlashSheridan) [inline lld/test/ELF/section-symbols.test:1]: This is also upstream, so it needs an explanation and perhaps a Matter Labs wrapper.  Have pity upon a poor Windows-phobic compiler engineer who might someday have to deal with this file.

File: lld/unittests/CMakeLists.txt
Comment (FlashSheridan) [inline lld/unittests/CMakeLists.txt:7]: * Please be especially careful about labelling our changes to upstream CMake files.  (Extra pity for potential future compiler engineers unfamiliar with both Windows and the intricacies of CMake…)
* Grammar nit (esp. for upstream)  s/b “Temporarily disable”.

**Review by FlashSheridan — CHANGES_REQUESTED** [review]

**Comment by antonbaliasnikov** [comment]
@FlashSheridan , thanks for your comments. As I was able to get a Windows host yesterday, I will try to compare it with the upstream and get better understanding of these failures. Putting this PR to `WIP` for now.

**Comment by antonbaliasnikov** [comment]
Closing this one with the proper disabling on our side: https://github.com/matter-labs/era-compiler-llvm/pull/702

**Pull request closed without merging**
